eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala

```diff
@@ -401,4 +428,35 @@ object Peer {
       globalFeatures = nodeParams.globalFeatures,
       localFeatures = nodeParams.localFeatures)
   }
+
+  /**
+    * filter out gossip messages using the provided origin and optional timestamp range
+    * @param rebroadcast rebroadcast message
+    * @param self messages which have been sent by `self` will be filtered out
+    * @param gossipTimestampFilter optional gossip timestamp range
+    * @return a filtered (channel announcements, channel updates, node announcements) tuple
+    */
+  def filterGossipMessages(rebroadcast: Rebroadcast, self: ActorRef, gossipTimestampFilter: Option[GossipTimestampFilter]): (Seq[ChannelAnnouncement], Seq[ChannelUpdate], Seq[NodeAnnouncement]) = {
+
+    // check if this message has a timestamp that matches our timestamp filter
+    def checkTimestamp(routingMessage: RoutingMessage): Boolean = gossipTimestampFilter match {
+      case None => true // no filtering
+      case Some(GossipTimestampFilter(_, firstTimestamp, timestampRange)) => routingMessage match {
+        case hts: HasTimestamp => hts.timestamp >= firstTimestamp && hts.timestamp <= firstTimestamp + timestampRange
+        case _ => true
+      }
+    }
+
+    // we filter out announcements that we received from this node, and we also filter out updates against their timestamp filter
+    val updates1 = rebroadcast.updates.collect { case (a, origins) if !origins.contains(self) && checkTimestamp(a) => a} toSeq
+    val nodes1 = rebroadcast.nodes.collect { case (a, origins) if !origins.contains(self) && checkTimestamp(a) => a } toSeq
```

Comment:
we should only send nodes that have a channel attached, filtering needs to happen later in the function